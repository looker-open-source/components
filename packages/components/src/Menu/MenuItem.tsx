/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import { CompatibleHTMLProps } from '@looker/design-tokens'
import { IconNames } from '@looker/icons'
import styled from 'styled-components'
import React, { FC, ReactNode, useContext, useState, useEffect } from 'react'
import { Icon } from '../Icon'
import { MenuContext, MenuItemContext } from './MenuContext'
import { MenuItemLayout, MenuItemLayoutGrid } from './MenuItemLayout'

export interface MenuItemProps extends CompatibleHTMLProps<HTMLElement> {
  compact?: boolean

  detail?: ReactNode
  icon?: IconNames
  /**
   * Indicates the MenuItem is checked
   */
  current?: boolean
  /**
   * Sets the correct accessible role for the MenuItem:
   * - Use **'link'** for items that navigation to another page
   * - Use **'button'** for items that trigger in page interactions, like displaying a dialog
   * @default 'button'
   *
   */
  itemRole?: 'link' | 'button'
}

const MenuItemInternal: FC<MenuItemProps> = (props) => {
  const {
    children,
    className,
    compact: propCompact,
    current,
    detail,
    disabled,
    href,
    icon,
    itemRole,
    onBlur,
    onClick,
    onKeyUp,
    target,
  } = props

  const [isFocusVisible, setFocusVisible] = useState(false)
  const { compact: contextCompact } = useContext(MenuItemContext)
  const compact = propCompact === undefined ? contextCompact : propCompact

  const handleOnKeyUp = (event: React.KeyboardEvent<HTMLLIElement>) => {
    setFocusVisible(true)
    onKeyUp && onKeyUp(event)
  }

  const handleOnBlur = (event: React.FocusEvent<HTMLLIElement>) => {
    setFocusVisible(false)
    onBlur && onBlur(event)
  }

  const { setOpen } = useContext(MenuContext)
  const handleOnClick = (event: React.MouseEvent<HTMLLIElement>) => {
    setFocusVisible(false)
    onClick && onClick(event)
    // Close the Menu (unless event has preventDefault in onClick)
    if (setOpen && !event.defaultPrevented) {
      setOpen(false)
    }
  }

  const { renderIconPlaceholder, setRenderIconPlaceholder } = useContext(
    MenuItemContext
  )

  useEffect(() => {
    icon && setRenderIconPlaceholder(true)
  }, [icon, setRenderIconPlaceholder])

  const renderedIcon = icon ? (
    <Icon
      name={icon}
      mr="xsmall"
      size={20 / (compact ? 1.25 : 1)}
      color="text6"
    />
  ) : (
    renderIconPlaceholder && <div />
  )

  const Component = itemRole === 'link' ? 'a' : 'button'

  return (
    <MenuItemLayout
      aria-current={current && 'true'}
      compact={compact}
      disabled={disabled}
      focusVisible={isFocusVisible}
      hasIcon={Boolean(renderedIcon)}
      onBlur={handleOnBlur}
      onClick={handleOnClick}
      onKeyUp={handleOnKeyUp}
      className={className}
    >
      <Component href={href} role="menuitem" target={target}>
        <MenuItemLayoutGrid>
          {renderedIcon}
          {children}
        </MenuItemLayoutGrid>
      </Component>
      {detail && <Detail>{detail}</Detail>}
    </MenuItemLayout>
  )
}

export const MenuItem = styled(MenuItemInternal)``

const Detail = styled.div`
  color: ${({ theme: { colors } }) => colors.text6};
  margin-left: auto;
  margin-right: ${({ theme: { space } }) => space.medium};
  padding-left: ${({ theme: { space } }) => space.large};
`
