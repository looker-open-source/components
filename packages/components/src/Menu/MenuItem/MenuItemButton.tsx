/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import styled from 'styled-components'
import {
  reset,
  space,
  CompatibleHTMLProps,
  SpaceProps,
} from '@looker/design-tokens'

export interface MenuItemButtonProps
  extends SpaceProps,
    CompatibleHTMLProps<HTMLElement> {}

export const MenuItemButton = styled.button<MenuItemButtonProps>`
  ${reset}
  ${space}

  align-items: center;
  background: transparent;
  border: 0;
  color: inherit;
  cursor: pointer;
  display: flex;
  font-size: inherit;
  flex: 1;
  outline: none;
  text-align: left;
  text-decoration: none;

  &:hover,
  &:focus {
    text-decoration: none;
    color: inherit;
  }
`
