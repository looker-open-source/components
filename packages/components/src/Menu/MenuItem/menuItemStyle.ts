/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import { FontWeights, FontSizes, theme } from '@looker/design-tokens'

const { colors } = theme

export interface MenuItemStateStyle {
  bg: string
  color: string
  fontWeight: FontWeights
  fontSize: FontSizes
  iconColor: string
  iconSize: number
}

export interface MenuItemStyle {
  current: MenuItemStateStyle
  hover: MenuItemStateStyle
  initial: MenuItemStateStyle
  marker: {
    color: string
    size: number
  }
}

const initial: MenuItemStateStyle = {
  bg: colors.background,
  color: colors.text3,
  fontSize: 'small',
  fontWeight: 'normal',
  iconColor: colors.text6,
  iconSize: 20,
}

const hover: MenuItemStateStyle = {
  ...initial,
  bg: colors.ui1,
  color: colors.text0,
}

const current: MenuItemStateStyle = {
  ...hover,
  fontWeight: 'bold',
  iconColor: colors.text0,
}

export const defaultMenuItemStyle: MenuItemStyle = {
  current,
  hover,
  initial,
  marker: {
    color: colors.text0,
    size: 4,
  },
}
