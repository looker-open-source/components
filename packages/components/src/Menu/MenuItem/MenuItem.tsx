/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import merge from 'lodash/merge'
import {
  CompatibleHTMLProps,
  ColorProps,
  FontSizes,
  SpaceProps,
  TypographyProps,
} from '@looker/design-tokens'
import { IconNames } from '@looker/icons'
import React, { FC, ReactNode, useContext, useState } from 'react'
import { Icon } from '../../Icon'
import { MenuItemStyleContext } from '../MenuContext'
import { MenuItemButton } from './MenuItemButton'
import {
  MenuItemCustomization,
  MenuItemStateCustomizations,
} from './menuItemCustomization'
import { MenuItemDetail } from './MenuItemDetail'
import { MenuItemListItem } from './MenuItemListItem'
import {
  defaultMenuItemStyle,
  MenuItemStateStyle,
  MenuItemStyle,
} from './menuItemStyle'

export interface MenuSharedProps {
  customizationProps?: MenuItemCustomization
  compact?: boolean
}

// For merging compact and customizationProps from props with those from context
export function useMenuItemStyleContext(props: MenuSharedProps) {
  const { customizationProps: propCustomizations, compact: compactProp } = props
  const {
    customizationProps: contextCustomizations,
    compact: contextCompact,
  } = useContext(MenuItemStyleContext)
  const parentCustomizations = contextCustomizations || {}

  let customizationProps = parentCustomizations || propCustomizations
  if (customizationProps && parentCustomizations) {
    customizationProps = merge({}, parentCustomizations, propCustomizations)
  }

  const compact = compactProp === undefined ? contextCompact : compactProp

  return { compact, customizationProps }
}

const assignCustomizations = (
  defaultStyle: MenuItemStyle,
  changes?: MenuItemCustomization
): MenuItemStyle => {
  const { bg, color, iconColor, fontWeight, fontSize, iconSize } =
    changes || ({} as MenuItemCustomization)

  const customMarker = changes ? changes.marker : {}

  // Need to spread fontSize & iconSize across all states
  const defaults = {
    fontSize: fontSize || ('small' as FontSizes),
    iconSize: iconSize || 20,
  }

  const base: MenuItemStateCustomizations = {}

  bg && (base.bg = bg)
  color && (base.color = color)
  fontWeight && (base.fontWeight = fontWeight)
  iconColor && (base.iconColor = iconColor)

  const initial: MenuItemStateStyle = {
    ...defaultStyle.initial,
    ...defaults,
    ...base,
  }

  const current: MenuItemStateStyle = {
    ...defaultStyle.current,
    ...defaults,
    ...base,
    ...(changes ? changes.current : {}),
  }

  const hover: MenuItemStateStyle = {
    ...defaultStyle.hover,
    ...defaults,
    ...base,
    ...(changes ? changes.hover : {}),
  }

  const marker = { ...defaultStyle.marker, ...customMarker }

  return {
    current,
    hover,
    initial,
    marker,
  }
}

export interface MenuItemProps
  extends CompatibleHTMLProps<HTMLElement>,
    ColorProps,
    SpaceProps,
    TypographyProps,
    MenuSharedProps {
  detail?: ReactNode
  icon?: IconNames
  /**
   * Indicates the MenuItem is checked
   */
  current?: boolean
  /**
   * Sets the correct accessible role for the MenuItem:
   * - Use **'link'** for items that navigation to another page
   * - Use **'button'** for items that trigger in page interactions, like displaying a modal
   * @default 'button'
   *
   */
  itemRole?: 'link' | 'button'
}

export const MenuItem: FC<MenuItemProps> = (props) => {
  const {
    children,
    compact: propCompact,
    customizationProps: propCustomizations,
    current,
    detail,
    href,
    icon,
    itemRole,
    onBlur,
    onClick,
    onKeyUp,
    target,
    ...remainingProps
  } = props

  const [isFocusVisible, setFocusVisible] = useState(false)

  const handleOnKeyUp = (event: React.KeyboardEvent<HTMLLIElement>) => {
    setFocusVisible(true)
    onKeyUp && onKeyUp(event)
  }

  const handleOnBlur = (event: React.FocusEvent<HTMLLIElement>) => {
    setFocusVisible(false)
    onBlur && onBlur(event)
  }

  const handleOnClick = (event: React.MouseEvent<HTMLLIElement>) => {
    setFocusVisible(false)
    onClick && onClick(event)
  }

  const { customizationProps, compact } = useMenuItemStyleContext({
    compact: propCompact,
    customizationProps: propCustomizations,
  })
  const compactIconModifier = compact ? 1.25 : 1

  const style = assignCustomizations(defaultMenuItemStyle, customizationProps)
  const styleState = current ? style.current : style.initial
  const { iconSize, iconColor, ...listItemProps } = styleState

  const { p, py, px, pr, pl, pt, pb, ...outerProps } = remainingProps

  const clickTargetProps = {
    p,
    pb: pb || py || p || compact ? 'xxsmall' : 'small',
    pl: pl || px || p || 'medium',
    pr: pr || px || p || 'medium',
    pt: pt || py || p || compact ? 'xxsmall' : 'small',
  }

  return (
    <MenuItemListItem
      aria-current={current && 'page'}
      onClick={handleOnClick}
      current={current}
      focusVisible={isFocusVisible}
      itemStyle={style}
      onKeyUp={handleOnKeyUp}
      onBlur={handleOnBlur}
      {...listItemProps}
      {...outerProps}
    >
      <MenuItemButton
        as={itemRole === 'link' ? 'a' : 'button'}
        href={href}
        role="menuitem"
        target={target}
        {...clickTargetProps}
      >
        {icon && (
          <Icon
            name={icon}
            mr="xsmall"
            size={iconSize / compactIconModifier}
            color={iconColor}
          />
        )}
        {children}
      </MenuItemButton>
      {detail && <MenuItemDetail>{detail}</MenuItemDetail>}
    </MenuItemListItem>
  )
}
