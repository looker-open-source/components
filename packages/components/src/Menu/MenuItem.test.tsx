/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import 'jest-styled-components'
import React from 'react'
import { renderWithTheme } from '@looker/components-test-utils'
import { screen, fireEvent } from '@testing-library/react'

import { MenuItem } from './MenuItem'

describe('MenuItem', () => {
  test('MenuItem renders', () => {
    renderWithTheme(<MenuItem>who!</MenuItem>)
    expect(screen.getByText('who!')).toBeVisible()
  })

  test('MenuItem - detail', () => {
    renderWithTheme(<MenuItem detail="Is an excellent question">who!</MenuItem>)
    expect(screen.getByText('Is an excellent question')).toBeVisible()
  })

  test('MenuItem - icon', () => {
    renderWithTheme(<MenuItem icon="Beaker">Icon</MenuItem>)
    expect(screen.getByText('Icon')).toBeVisible()
  })

  test('MenuItem - artwork', () => {
    renderWithTheme(
      <MenuItem
        iconArtwork={
          <svg xmlns="http://www.w3.org/2000/svg">
            <title>SVG Title Here</title>
          </svg>
        }
      >
        Artwork
      </MenuItem>
    )
    expect(screen.getByTitle('SVG Title Here')).toBeInTheDocument()
  })

  test('MenuItem - disabled to be a button', () => {
    const callbackFn = jest.fn()

    renderWithTheme(
      <MenuItem
        disabled={true}
        href="https://google.com"
        onClick={callbackFn}
        target="_blank"
      >
        Item
      </MenuItem>
    )
    const item = screen.getByText('Item')
    expect(item.closest('button')).toBeInTheDocument()
  })

  test('MenuItem - disabled is not clickable', () => {
    const callbackFn = jest.fn()

    renderWithTheme(
      <MenuItem itemRole="button" disabled onClick={callbackFn}>
        Item
      </MenuItem>
    )

    const item = screen.getByText('Item')
    fireEvent.click(item)

    expect(callbackFn).toHaveBeenCalledTimes(0)
  })
})
