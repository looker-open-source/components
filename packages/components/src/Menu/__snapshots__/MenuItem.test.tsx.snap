// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuItem - current 1`] = `
.c3 {
  width: 20px;
  height: 20px;
  margin-right: 0.5rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  color: #C1C6CC;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c3 svg {
  height: 100%;
  width: 100%;
}

.c4 button .c1,
.c4 a .c1 {
  display: grid;
  grid-gap: 0.5rem;
  grid-template-columns: 1fr;
}

.c4 .c2 {
  color: #C1C6CC;
  -webkit-transition: color 150ms cubic-bezier(0.86,0,0.07,1);
  transition: color 150ms cubic-bezier(0.86,0,0.07,1);
}

.c4:hover .c2,
.c4[aria-current='true'] .c2 {
  color: #262D33;
}

.c4[disabled]:hover .c2 {
  color: #C1C6CC;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #262D33;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  font-size: 0.875rem;
  font-weight: 400;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background 150ms cubic-bezier(0.86,0,0.07,1), color 150ms cubic-bezier(0.86,0,0.07,1);
  transition: background 150ms cubic-bezier(0.86,0,0.07,1), color 150ms cubic-bezier(0.86,0,0.07,1);
}

.c0 button,
.c0 a {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: transparent;
  border: none;
  color: inherit;
  cursor: pointer;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: inherit;
  font-weight: inherit;
  outline: none;
  padding: 0.75rem 1rem;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 button:hover,
.c0 a:hover,
.c0 button:focus,
.c0 a:focus {
  color: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 button .c1,
.c0 a .c1 {
  display: grid;
  grid-gap: 0.5rem;
  grid-template-columns: 20px 1fr;
}

.c0 .c2 {
  color: #C1C6CC;
  -webkit-transition: color 150ms cubic-bezier(0.86,0,0.07,1);
  transition: color 150ms cubic-bezier(0.86,0,0.07,1);
}

.c0:hover,
.c0[aria-current='true'] {
  background: #F5F6F7;
  color: #262D33;
}

.c0:hover .c2,
.c0[aria-current='true'] .c2 {
  color: #262D33;
}

.c0[aria-current='true'] {
  font-weight: 600;
}

.c0[disabled] {
  color: #939BA5;
}

.c0[disabled] button,
.c0[disabled] a {
  cursor: not-allowed;
}

.c0[disabled]:hover {
  background: #FFFFFF;
  color: #939BA5;
}

.c0[disabled]:hover .c2 {
  color: #C1C6CC;
}

<li
  aria-current="true"
  className="c0 "
  onBlur={[Function]}
  onClick={[Function]}
  onKeyUp={[Function]}
>
  <button
    role="menuitem"
  >
    <div
      className="c1 "
    >
      <div
        className="c2 c3"
      >
        <svg
          fill="currentColor"
          height="100%"
          viewBox="0 0 24 24"
          width="100%"
        >
          <title>
            Home
          </title>
          <path
            d="M12 5.69l5 4.5V18h-2v-6H9v6H7V10.19l5-4.5zM12 3L2 12h3v8h6v-6h2v6h6v-8h3L12 3z"
            fill="currentColor"
          />
        </svg>
      </div>
      who!
    </div>
  </button>
</li>
`;

exports[`MenuItem - detail 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #262D33;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  font-size: 0.875rem;
  font-weight: 400;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background 150ms cubic-bezier(0.86,0,0.07,1), color 150ms cubic-bezier(0.86,0,0.07,1);
  transition: background 150ms cubic-bezier(0.86,0,0.07,1), color 150ms cubic-bezier(0.86,0,0.07,1);
}

.c0 button,
.c0 a {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: transparent;
  border: none;
  color: inherit;
  cursor: pointer;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: inherit;
  font-weight: inherit;
  outline: none;
  padding: 0.75rem 1rem;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 button:hover,
.c0 a:hover,
.c0 button:focus,
.c0 a:focus {
  color: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 button .c1,
.c0 a .c1 {
  display: grid;
  grid-gap: 0.5rem;
  grid-template-columns: 1fr;
}

.c0 .c4 {
  color: #C1C6CC;
  -webkit-transition: color 150ms cubic-bezier(0.86,0,0.07,1);
  transition: color 150ms cubic-bezier(0.86,0,0.07,1);
}

.c0:hover,
.c0[aria-current='true'] {
  background: #F5F6F7;
  color: #262D33;
}

.c0:hover .c4,
.c0[aria-current='true'] .c4 {
  color: #262D33;
}

.c0[aria-current='true'] {
  font-weight: 600;
}

.c0[disabled] {
  color: #939BA5;
}

.c0[disabled] button,
.c0[disabled] a {
  cursor: not-allowed;
}

.c0[disabled]:hover {
  background: #FFFFFF;
  color: #939BA5;
}

.c0[disabled]:hover .c4 {
  color: #C1C6CC;
}

.c3 button .c1,
.c3 a .c1 {
  display: grid;
  grid-gap: 0.5rem;
  grid-template-columns: 20px 1fr;
}

.c2 {
  color: #C1C6CC;
  margin-left: auto;
  margin-right: 1rem;
  padding-left: 1.25rem;
}

<li
  className="c0 "
  onBlur={[Function]}
  onClick={[Function]}
  onKeyUp={[Function]}
>
  <button
    role="menuitem"
  >
    <div
      className="c1 "
    >
      who!
    </div>
  </button>
  <div
    className="c2"
  >
    Is an excellent question
  </div>
</li>
`;

exports[`MenuItem - icon 1`] = `
.c3 {
  width: 20px;
  height: 20px;
  margin-right: 0.5rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  color: #C1C6CC;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c3 svg {
  height: 100%;
  width: 100%;
}

.c4 button .c1,
.c4 a .c1 {
  display: grid;
  grid-gap: 0.5rem;
  grid-template-columns: 1fr;
}

.c4 .c2 {
  color: #C1C6CC;
  -webkit-transition: color 150ms cubic-bezier(0.86,0,0.07,1);
  transition: color 150ms cubic-bezier(0.86,0,0.07,1);
}

.c4:hover .c2,
.c4[aria-current='true'] .c2 {
  color: #262D33;
}

.c4[disabled]:hover .c2 {
  color: #C1C6CC;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #262D33;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  font-size: 0.875rem;
  font-weight: 400;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background 150ms cubic-bezier(0.86,0,0.07,1), color 150ms cubic-bezier(0.86,0,0.07,1);
  transition: background 150ms cubic-bezier(0.86,0,0.07,1), color 150ms cubic-bezier(0.86,0,0.07,1);
}

.c0 button,
.c0 a {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: transparent;
  border: none;
  color: inherit;
  cursor: pointer;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: inherit;
  font-weight: inherit;
  outline: none;
  padding: 0.75rem 1rem;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 button:hover,
.c0 a:hover,
.c0 button:focus,
.c0 a:focus {
  color: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 button .c1,
.c0 a .c1 {
  display: grid;
  grid-gap: 0.5rem;
  grid-template-columns: 20px 1fr;
}

.c0 .c2 {
  color: #C1C6CC;
  -webkit-transition: color 150ms cubic-bezier(0.86,0,0.07,1);
  transition: color 150ms cubic-bezier(0.86,0,0.07,1);
}

.c0:hover,
.c0[aria-current='true'] {
  background: #F5F6F7;
  color: #262D33;
}

.c0:hover .c2,
.c0[aria-current='true'] .c2 {
  color: #262D33;
}

.c0[aria-current='true'] {
  font-weight: 600;
}

.c0[disabled] {
  color: #939BA5;
}

.c0[disabled] button,
.c0[disabled] a {
  cursor: not-allowed;
}

.c0[disabled]:hover {
  background: #FFFFFF;
  color: #939BA5;
}

.c0[disabled]:hover .c2 {
  color: #C1C6CC;
}

<li
  className="c0 "
  onBlur={[Function]}
  onClick={[Function]}
  onKeyUp={[Function]}
>
  <button
    role="menuitem"
  >
    <div
      className="c1 "
    >
      <div
        className="c2 c3"
      >
        <svg
          fill="currentColor"
          height="100%"
          viewBox="0 0 24 24"
          width="100%"
        >
          <title>
            Beaker
          </title>
          <path
            d="M15.75 4.502a1.249 1.249 0 100-2.5 1.25 1.25 0 100 2.5zM12 5.752a1.25 1.25 0 10-.002-2.501A1.25 1.25 0 0012 5.753zm7.22 15.01c-1.254-2.415-3.47-7.511-3.47-7.511v-5h.625a.626.626 0 000-1.25h-7.5a.626.626 0 000 1.25H9.5v5s-2.168 5.276-3.393 7.633c-.58 1.118-.58 1.118.803 1.118h11.747c.56 0 1.21 0 .565-1.24h-.001zm-8.47-7.51v-5h3.75v5l.535 1.25h-4.793c.3-.733.508-1.25.508-1.25zm6.172 7.5h-8.33c-1.153 0-1.153 0-.67-.957.513-1.015 1.223-2.658 1.803-4.043h5.855c.59 1.338 1.3 2.917 1.814 3.935.536 1.064-.005 1.064-.471 1.064z"
            fill="currentColor"
          />
        </svg>
      </div>
      who!
    </div>
  </button>
</li>
`;

exports[`MenuItem 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #262D33;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  font-size: 0.875rem;
  font-weight: 400;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background 150ms cubic-bezier(0.86,0,0.07,1), color 150ms cubic-bezier(0.86,0,0.07,1);
  transition: background 150ms cubic-bezier(0.86,0,0.07,1), color 150ms cubic-bezier(0.86,0,0.07,1);
}

.c0 button,
.c0 a {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: transparent;
  border: none;
  color: inherit;
  cursor: pointer;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: inherit;
  font-weight: inherit;
  outline: none;
  padding: 0.75rem 1rem;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 button:hover,
.c0 a:hover,
.c0 button:focus,
.c0 a:focus {
  color: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 button .c1,
.c0 a .c1 {
  display: grid;
  grid-gap: 0.5rem;
  grid-template-columns: 1fr;
}

.c0 .c2 {
  color: #C1C6CC;
  -webkit-transition: color 150ms cubic-bezier(0.86,0,0.07,1);
  transition: color 150ms cubic-bezier(0.86,0,0.07,1);
}

.c0:hover,
.c0[aria-current='true'] {
  background: #F5F6F7;
  color: #262D33;
}

.c0:hover .c2,
.c0[aria-current='true'] .c2 {
  color: #262D33;
}

.c0[aria-current='true'] {
  font-weight: 600;
}

.c0[disabled] {
  color: #939BA5;
}

.c0[disabled] button,
.c0[disabled] a {
  cursor: not-allowed;
}

.c0[disabled]:hover {
  background: #FFFFFF;
  color: #939BA5;
}

.c0[disabled]:hover .c2 {
  color: #C1C6CC;
}

<li
  className="c0 "
  onBlur={[Function]}
  onClick={[Function]}
  onKeyUp={[Function]}
>
  <button
    role="menuitem"
  >
    <div
      className="c1 "
    >
      who!
    </div>
  </button>
</li>
`;
